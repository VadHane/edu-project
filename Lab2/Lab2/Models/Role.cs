using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace Lab2.Models
{
    public class Role
    {
        public Role()
        {
            Users = new HashSet<User>();
        }

        public Role(Guid id)
        {
            Id = id;
            Users = new HashSet<User>();
        }

        public Guid Id { get; set; }

        
        [MaxLength(20)]
        public string Name { get; set; }

        public virtual ICollection<User> Users { get; set; }
    }
}
