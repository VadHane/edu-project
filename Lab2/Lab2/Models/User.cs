using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace Lab2.Models
{
    public class User
    {
        public User()
        {
            Roles = new List<Role>();
        }

        public Guid Id { get; set; }

        [Required]
        [MaxLength(30)]
        public string Email { get; set; }

        [Required]
        [MaxLength(15)]
        public string FirstName { get; set; }

        [MaxLength(20)]
        public string LastName { get; set; }
        public string ImageBlobKey { get; set; }


        public virtual ICollection<Role> Roles { get; set; }
    }
}
