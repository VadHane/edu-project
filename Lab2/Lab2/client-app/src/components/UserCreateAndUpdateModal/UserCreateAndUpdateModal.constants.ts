export const FIRST_NAME_PLACEHOLDER = 'First name';
export const LAST_NAME_PLACEHOLDER = 'Last name';
export const EMAIL_PLACEHOLDER = 'email';
