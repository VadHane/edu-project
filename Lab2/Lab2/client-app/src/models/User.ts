import {Role} from './Role';

export interface User {
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    imageBlobKey: string,
    roles: Array<Role>,
};
