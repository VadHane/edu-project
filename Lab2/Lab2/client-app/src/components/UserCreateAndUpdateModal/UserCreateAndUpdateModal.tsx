/* eslint-disable react-hooks/rules-of-hooks */
import React, { FunctionComponent, useEffect, useState } from 'react';
import { NavLink, useNavigate, useParams } from 'react-router-dom';
import { Role } from '../../models/Role';
import AddedRolesList from './AddedRolesList';
import AvailableRolesList from './AvailableRolesList';
import UserCreateAndUpdateModalProps from './UserCreateAndUpdateModal.types';
import { User } from '../../models/User';
import {
    FILE_NOT_IMAGE_EXCEPTION,
    INCORECT_USER_ID_EXCEPTION,
    INCORRECT_EMAIL_EXCEPTION,
    LENGTH_OF_NAME_EXCEPTION,
    LENGTH_OF_SURNAME_EXCEPTION,
    TRY_AGAIN_LATER_EXCEPTION,
} from '../../App.constants';
import './UserCreateAndUpdateModal.css';
import UserWarningModal from '../UserWarningModal';
import {
    EMAIL_PLACEHOLDER,
    FIRST_NAME_PLACEHOLDER,
    LAST_NAME_PLACEHOLDER,
} from './UserCreateAndUpdateModal.constants';

const UserCreateAndUpdateModal: FunctionComponent<UserCreateAndUpdateModalProps> = (
    props: UserCreateAndUpdateModalProps,
) => {
    const { id } = useParams();
    const foundUser = props.getUserById(id);

    if (!foundUser) {
        return <UserWarningModal message={INCORECT_USER_ID_EXCEPTION} />;
    }

    const navigate = useNavigate();
    const [firstName, setFirstName] = useState<string>(foundUser.firstName);
    const [lastName, setLastName] = useState<string>(foundUser.lastName);
    const [email, setEmail] = useState<string>(foundUser.email);
    const [addedRoles, setAddedRoles] = useState<Array<Role>>(foundUser.roles);
    const [availableRoles, setAvailableRoles] = useState<Array<Role>>([]);
    const [exceptionMessage, setExceptionMessage] = useState<string>('');

    const file = React.createRef<HTMLInputElement>();

    useEffect(() => {
        props
            .getAllRolesAsync()
            .then((roles: Array<Role>) => setAvailableRoles([...roles]));
    }, [props]);

    useEffect(() => {
        setExceptionMessage('');
    }, [firstName, lastName, email]);

    const backgroundNode: React.ReactNode = (
        <NavLink to={'/'} title="Close modal window">
            <div className="background"></div>
        </NavLink>
    );

    const exceptionString: React.ReactNode = (
        <div className="exception-message">
            <span>{exceptionMessage}</span>
        </div>
    );

    const modalWindowNode: React.ReactNode = (
        <div className="window">
            <form>
                <div className="formRows">
                    <br />
                    {exceptionString}
                    <input
                        type="text"
                        name="first_name"
                        id="first_name"
                        placeholder={FIRST_NAME_PLACEHOLDER}
                        value={firstName}
                        onChange={(event) => {
                            setFirstName(event.target.value);
                        }}
                    />
                    <br />

                    <input
                        type="text"
                        name="last_name"
                        id="last_name"
                        placeholder={LAST_NAME_PLACEHOLDER}
                        value={lastName}
                        onChange={(event) => {
                            setLastName(event.target.value);
                        }}
                    />
                    <br />

                    <input
                        type="email"
                        name="email"
                        id="email"
                        placeholder={EMAIL_PLACEHOLDER}
                        value={email}
                        onChange={(event) => {
                            setEmail(event.target.value);
                        }}
                    />
                    <br />

                    <AddedRolesList
                        roles={addedRoles}
                        removeAddedRole={(role: Role) => addAvailableRole(role)}
                    />

                    <AvailableRolesList
                        roles={availableRoles}
                        addedRoles={addedRoles}
                        addRole={(role: Role) => addRole(role)}
                        createNewRole={(role: Role) => createNewRole(role)}
                    />

                    <input type="file" accept="image/*" ref={file} />
                </div>
            </form>
            <button onClick={() => onSubmitAction()}>{props.buttonContent}</button>
        </div>
    );

    const addRole = (role: Role): void => {
        setAvailableRoles((currentValue: Array<Role>) => {
            const indexOfSelectedRole = currentValue?.findIndex(
                (_role: Role) => _role.id === role.id,
            );
            const prevList = currentValue.slice(0, indexOfSelectedRole);
            const endList = currentValue.slice(
                indexOfSelectedRole + 1,
                currentValue.length,
            );

            return [...prevList, ...endList];
        });
        setAddedRoles((currentValue) => [...currentValue, role]);
    };

    const addAvailableRole = (role: Role) => {
        setAddedRoles((currentValue: Array<Role>) => {
            const indexOfSelectedRole = currentValue?.findIndex(
                (_role: Role) => _role.id === role.id,
            );
            const prevList = currentValue.slice(0, indexOfSelectedRole);
            const endList = currentValue.slice(
                indexOfSelectedRole + 1,
                currentValue.length,
            );

            return [...prevList, ...endList];
        });
        setAvailableRoles((currentValue) => [...currentValue, role]);
    };

    const createNewRole = (role: Role): void => {
        props.createNewRole(role).then((role: Role) => {
            setAvailableRoles((currentValue: Array<Role>) => [...currentValue, role]);
        });
    };

    const validateInputFields = (): boolean => {
        if (firstName.trim().length < 3) {
            setExceptionMessage(LENGTH_OF_NAME_EXCEPTION);
            return false;
        }

        if (lastName.trim().length < 3) {
            setExceptionMessage(LENGTH_OF_SURNAME_EXCEPTION);
            return false;
        }

        const pattern = /^[^ ]+@[^ ]+\.[a-z]{2,3}$/;
        if (!email.match(pattern)) {
            setExceptionMessage(INCORRECT_EMAIL_EXCEPTION);
            return false;
        }

        if (file.current?.files?.item(0)?.type.split('/')[0] !== 'image') {
            setExceptionMessage(FILE_NOT_IMAGE_EXCEPTION);
            return false;
        }

        return true;
    };

    const onSubmitAction = (): void => {
        if (!validateInputFields()) return;

        const user: User = {
            id: foundUser.id,
            firstName: firstName,
            lastName: lastName,
            email: email,
            imageBlobKey: foundUser.imageBlobKey,
            roles: addedRoles,
        };

        const userPhoto = file.current?.files?.item(0);
        props.resultActionAsync(user, userPhoto).then((done: Boolean) => {
            if (done) {
                navigate('/');
            } else {
                setExceptionMessage(TRY_AGAIN_LATER_EXCEPTION);
            }
        });
    };

    return (
        <div className="wrapper">
            {backgroundNode}
            {modalWindowNode}
        </div>
    );
};

export default UserCreateAndUpdateModal;
