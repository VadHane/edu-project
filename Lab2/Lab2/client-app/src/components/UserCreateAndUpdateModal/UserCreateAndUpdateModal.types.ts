import { Role } from '../../models/Role';
import { User } from '../../models/User';

type UserCreateAndUpdateModalProps = {
    buttonContent: string;
    getAllRolesAsync: () => Promise<Array<Role>>;
    getUserById: (id?: string) => User | undefined;
    createNewRole: (role: Role) => Promise<Role>;

    // * Handler for click action for the button of form.
    // ! It can be add or edit operation.
    resultActionAsync: (user: User, file: any) => Promise<boolean>;
};

export default UserCreateAndUpdateModalProps;
