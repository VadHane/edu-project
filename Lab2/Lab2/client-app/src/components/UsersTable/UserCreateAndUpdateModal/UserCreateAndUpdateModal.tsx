import React, {FunctionComponent, useEffect, useState} from 'react';
import {NavLink, useNavigate} from 'react-router-dom';
import {Role} from '../../../models/Role';
import AddedRolesList from './AddedRolesList/AddedRolesList';
import AvailableRolesList from './AvailableRolesList/AvailableRolesList';
import UserCreateAndUpdateModalProps from './UserCreateAndUpdateModal.types';
import './UserCreateAndUpdateModal.css';
import {User} from '../../../models/User';

const UserCreateAndUpdateModal:
FunctionComponent<UserCreateAndUpdateModalProps> =
(props: UserCreateAndUpdateModalProps) => {
  const [firstName, setFirstName] = useState<string>(props.user.firstName);
  const [lastName, setLastName] = useState<string>(props.user.lastName);
  const [email, setEmail] = useState<string>(props.user.email);
  const [addedRoles, setAddedRoles] = useState<Array<Role>>(props.user.roles);
  const [availableRoles, setAvailableRoles] = useState<Array<Role>>([]);
  const [exceptionMessage, setExceptionMessage] = useState<string>('');
  const navigate = useNavigate();

  const file = React.createRef<HTMLInputElement>();

  useEffect(() => {
    props.getAllRolesAsync().then((roles: Array<Role>) =>
      setAvailableRoles([...roles]));
  }, []);

  useEffect(() => {
    setExceptionMessage('');
  }, [firstName, lastName, email]);

  const backgroundNode: React.ReactNode = (
    <NavLink to={'/'} title='Close modal window'>
      <div className="background"></div>
    </NavLink>
  );

  const exceptionString: React.ReactNode = (
    <div className='exceptionMessage'>
      <span>{exceptionMessage}</span>
    </div>
  );

  const modalWindowNode: React.ReactNode = (
    <div className="window">
      <form>
        <div className="formRows">
          <br/>
          {exceptionString}
          <input type="text"
            name='title'
            id='title'
            placeholder='First name'
            value={firstName}
            onChange={(event) => {
              setFirstName(event.target.value);
            }}/><br/>

          <input type="text"
            name='title'
            id='title'
            placeholder='Last name'
            value={lastName}
            onChange={(event) => {
              setLastName(event.target.value);
            }}/><br/>

          <input type="email"
            name='title'
            id='title'
            placeholder='email'
            value={email}
            onChange={(event) => {
              setEmail(event.target.value);
            }}/><br/>

          <AddedRolesList
            roles={addedRoles}
            removeAddedRole={(role: Role) => addAvailableRole(role)} />

          <AvailableRolesList
            roles={availableRoles}
            addedRoles={addedRoles}
            addRole={(role: Role) => addRole(role)}
            createNewRole={(role: Role) => createNewRole(role)}/>

          <input type="file"
            accept='image/*'
            ref={file}/>
        </div>
      </form>
      <button onClick={() => createNewUser()}>
        {props.btnCaption}
      </button>
    </div>
  );

  const addRole = (role: Role): void => {
    setAvailableRoles((currentValue: Array<Role>) => {
      const indexOfSelectedRole = currentValue?.findIndex(
          (_role: Role) => _role.id === role.id);
      const prevList = currentValue.slice(0, indexOfSelectedRole);
      const endList = currentValue.slice(
          indexOfSelectedRole+1, currentValue.length);

      return [...prevList, ...endList];
    });
    setAddedRoles((currentValue) => [...currentValue, role]);
  };

  const addAvailableRole = (role: Role) => {
    setAddedRoles((currentValue: Array<Role>) => {
      const indexOfSelectedRole = currentValue?.findIndex(
          (_role: Role) => _role.id === role.id);
      const prevList = currentValue.slice(0, indexOfSelectedRole);
      const endList = currentValue.slice(
          indexOfSelectedRole+1, currentValue.length);

      return [...prevList, ...endList];
    });
    setAvailableRoles((currentValue) => [...currentValue, role]);
  };

  const createNewRole = (role: Role): void => {
    props.createNewRole(role).then((role: Role) => {
      setAvailableRoles((currentValue: Array<Role>) =>
        [...currentValue, role]);
    });
  };

  const validateInputFields = (): boolean => {
    if (firstName.trim().length < 3) {
      setExceptionMessage('Length of username can not be less than 3 symbols.');
      return false;
    }

    if (lastName.trim().length < 3) {
      setExceptionMessage('Length of surname can not be less than 3 symbols.');
      return false;
    }

    if (email.trim().length < 3) {
      setExceptionMessage('Length of email can not be less than 3 symbols.');
      return false;
    }

    const pattern = /^[^ ]+@[^ ]+\.[a-z]{2,3}$/;
    if (!email.match(pattern)) {
      setExceptionMessage('Incorrect email.');
      return false;
    }

    if (file.current?.files?.item(0)?.type.split('/')[0] !== 'image') {
      setExceptionMessage('The file must be image.');
      return false;
    }

    return true;
  };

  const createNewUser = (): void => {
    if (!validateInputFields()) return;

    const newUser: User = {
      id: '',
      firstName: firstName,
      lastName: lastName,
      email: email,
      imageBlobKey: props.user.imageBlobKey,
      roles: addedRoles,
    };

    const userPhoto = file.current?.files?.item(0);

    props.resultActionAsync(newUser, userPhoto).then((done: Boolean) => {
      if (done) {
        navigate('/');
      } else {
        setExceptionMessage('Try again later.');
      }
    });
  };

  return (
    <div className="wrapper">
      {backgroundNode}
      {modalWindowNode}
    </div>
  );
};

export default UserCreateAndUpdateModal;
