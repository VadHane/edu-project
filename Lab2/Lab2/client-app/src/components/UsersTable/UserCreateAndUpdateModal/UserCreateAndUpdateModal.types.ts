import { Role } from '../../../models/Role';
import { User } from '../../../models/User';

type UserCreateAndUpdateModalProps = {
    user: User;
    btnCaption: string;
    getAllRolesAsync: () => Promise<Array<Role>>;
    createNewRole: (role: Role) => Promise<Role>;

    // * Handler for click action for the button of form.
    // ! It can be add or edit operation.
    resultActionAsync: (user: User, file: any) => Promise<boolean>;
};

export default UserCreateAndUpdateModalProps;
