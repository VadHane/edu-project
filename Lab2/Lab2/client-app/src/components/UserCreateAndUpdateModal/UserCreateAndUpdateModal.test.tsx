import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import {
    FILE_NOT_IMAGE_EXCEPTION,
    INCORECT_USER_ID_EXCEPTION,
    INCORRECT_EMAIL_EXCEPTION,
    LENGTH_OF_NAME_EXCEPTION,
    LENGTH_OF_SURNAME_EXCEPTION,
} from '../../App.constants';
import { Role } from '../../models/Role';
import { User } from '../../models/User';
import UserCreateAndUpdateModal from './UserCreateAndUpdateModal';

jest.mock('react-router-dom', () => ({
    ...(jest.requireActual('react-router-dom') as any),
    useNavigate: () => jest.fn(),
    useParams: () => jest.fn(),
    NavLink: jest.fn(() => <></>),
}));

jest.mock('./AddedRolesList');
jest.mock('./AvailableRolesList');

const testButtonCaption = 'test';
const testRolesList: Array<Role> = [
    {
        id: '1',
        name: 'test1',
    },
];
const testUser: User = {
    id: '1234',
    firstName: 'TestName',
    lastName: 'TestSurname',
    email: 'email@gmail.com',
    imageBlobKey: '/',
    roles: [],
};
const testEmptyUser: User = {
    id: '',
    firstName: '',
    lastName: '',
    email: '',
    imageBlobKey: null,
    roles: [],
};

const mockedGetAllRoles = jest.fn();
const mockedGetUserById = jest.fn();
const mockedCreateNewRole = jest.fn();
const mockedResultActionAsync = jest.fn();

const testModalWindowNode = (
    <UserCreateAndUpdateModal
        buttonContent={testButtonCaption}
        getAllRolesAsync={mockedGetAllRoles}
        getUserById={mockedGetUserById}
        createNewRole={mockedCreateNewRole}
        resultActionAsync={mockedResultActionAsync}
    />
);

describe('Test create/update modal window.', () => {
    beforeEach(() => {
        mockedGetUserById.mockReturnValue(testUser);
        mockedGetAllRoles.mockReturnValue(new Promise<Array<Role>>(() => testRolesList));
    });

    afterEach(cleanup);

    test('Component should render without errors.', () => {
        render(<>{testModalWindowNode}</>);

        expect(screen.getAllByRole('textbox').length).toBe(3);
        expect(
            screen.getByRole('button', { name: testButtonCaption }),
        ).toBeInTheDocument();
    });

    test('Component should render correct without data.', () => {
        const emptySting = '';

        render(
            <UserCreateAndUpdateModal
                buttonContent={emptySting}
                getAllRolesAsync={mockedGetAllRoles}
                getUserById={mockedGetUserById}
                createNewRole={mockedCreateNewRole}
                resultActionAsync={mockedResultActionAsync}
            />,
        );

        expect(screen.getAllByRole('textbox').length).toBe(3);
        expect(screen.getByRole('button', { name: emptySting })).toBeInTheDocument();
    });

    test('Component renders warning window, if getUserById returns undefined.', () => {
        mockedGetUserById.mockReturnValue(undefined);

        render(<>{testModalWindowNode}</>);

        expect(screen.queryAllByRole('textbox').length).toBe(0);
        expect(screen.queryByRole('button', { name: testButtonCaption })).toBeNull();
        expect(screen.getByText(INCORECT_USER_ID_EXCEPTION)).toBeInTheDocument();
    });

    test('When you click the "Approve" button, the component should contain a warning message if the name field is empty.', () => {
        mockedGetUserById.mockReturnValue(testEmptyUser);
        render(<>{testModalWindowNode}</>);

        fireEvent.click(screen.getByRole('button', { name: testButtonCaption }));

        expect(screen.getByText(LENGTH_OF_NAME_EXCEPTION)).toBeInTheDocument();
        expect(mockedResultActionAsync).toBeCalledTimes(0);
    });

    test('When you click the "Approve" button, the component should contain a warning message if the surname field is empty.', () => {
        const testInputName = 'test';

        mockedGetUserById.mockReturnValue(testEmptyUser);
        render(<>{testModalWindowNode}</>);

        fireEvent.change(screen.getAllByRole('textbox')[0], {
            target: { value: testInputName },
        });
        fireEvent.click(screen.getByRole('button', { name: testButtonCaption }));

        expect(screen.getByText(LENGTH_OF_SURNAME_EXCEPTION)).toBeInTheDocument();
        expect(mockedResultActionAsync).toBeCalledTimes(0);
    });

    test('When you click the "Approve" button, the component should contain a warning message if the email field is empty or incorrect.', () => {
        const testInputName = 'test';
        const testInputSurname = 'test';

        mockedGetUserById.mockReturnValue(testEmptyUser);
        render(<>{testModalWindowNode}</>);

        fireEvent.change(screen.getAllByRole('textbox')[0], {
            target: { value: testInputName },
        });
        fireEvent.change(screen.getAllByRole('textbox')[1], {
            target: { value: testInputSurname },
        });

        fireEvent.click(screen.getByRole('button', { name: testButtonCaption }));

        expect(screen.getByText(INCORRECT_EMAIL_EXCEPTION)).toBeInTheDocument();
        expect(mockedResultActionAsync).toBeCalledTimes(0);
    });

    test('When you click the "Approve" button, the component should contain a warning message if the file field is empty or file is not a image.', () => {
        const testInputName = 'test';
        const testInputSurname = 'test';
        const testEmail = 'test@test.ts';

        mockedGetUserById.mockReturnValue(testEmptyUser);
        render(<>{testModalWindowNode}</>);

        fireEvent.change(screen.getAllByRole('textbox')[0], {
            target: { value: testInputName },
        });
        fireEvent.change(screen.getAllByRole('textbox')[1], {
            target: { value: testInputSurname },
        });
        fireEvent.change(screen.getAllByRole('textbox')[2], {
            target: { value: testEmail },
        });

        fireEvent.click(screen.getByRole('button', { name: testButtonCaption }));

        expect(screen.getByText(FILE_NOT_IMAGE_EXCEPTION)).toBeInTheDocument();
        expect(mockedResultActionAsync).toBeCalledTimes(0);
    });

    test('Snapshot.', () => {
        const utils = render(<>{testModalWindowNode}</>);

        expect(utils).toMatchSnapshot();
    });
});
