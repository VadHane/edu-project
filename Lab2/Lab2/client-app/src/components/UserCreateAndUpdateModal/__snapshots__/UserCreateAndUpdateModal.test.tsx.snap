// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test create/update modal window. Snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="wrapper"
      >
        <div
          class="window"
        >
          <form>
            <div
              class="formRows"
            >
              <br />
              <div
                class="exception-message"
              >
                <span />
              </div>
              <input
                id="first_name"
                name="first_name"
                placeholder="First name"
                type="text"
                value="TestName"
              />
              <br />
              <input
                id="last_name"
                name="last_name"
                placeholder="Last name"
                type="text"
                value="TestSurname"
              />
              <br />
              <input
                id="email"
                name="email"
                placeholder="email"
                type="email"
                value="email@gmail.com"
              />
              <br />
              <input
                accept="image/*"
                type="file"
              />
            </div>
          </form>
          <button>
            test
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="wrapper"
    >
      <div
        class="window"
      >
        <form>
          <div
            class="formRows"
          >
            <br />
            <div
              class="exception-message"
            >
              <span />
            </div>
            <input
              id="first_name"
              name="first_name"
              placeholder="First name"
              type="text"
              value="TestName"
            />
            <br />
            <input
              id="last_name"
              name="last_name"
              placeholder="Last name"
              type="text"
              value="TestSurname"
            />
            <br />
            <input
              id="email"
              name="email"
              placeholder="email"
              type="email"
              value="email@gmail.com"
            />
            <br />
            <input
              accept="image/*"
              type="file"
            />
          </div>
        </form>
        <button>
          test
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
